homeassistant/components/camera/__init__.py

```diff
@@ -0,0 +1,1133 @@
+            str(time.time()))
+        attr['stream_url'] = '/api/camera_proxy_stream/' + self.entity_id
+        motion_time = self._last_motion_detected.strftime('%Y-%m-%d %H-%M-%S')
+        attr['last_motion_time'] = motion_time
+        attr['last_connected_address'] = self.last_connected_address
+        attr['child_entities'] = self._action_entities
+        attr['is_recording'] = self._is_recording
+        attr['is_taking_snapshot'] = self.is_taking_snapshot
+
+        attr.update(self.function_attributes)
+
+        return attr
+
+    @property
+    def still_image_url(self):
+        """ This should be implemented by different camera models. """
+        return self.base_url + self._still_image_url
+
+    def enable_motion_detection(self):
+        """ Enable the motion detection settings for the camera.
+            This should be overridden in derived classes """
+        if not self.is_motion_detection_supported:
+            return False
+        if self.is_motion_detection_enabled:
+            return True
+
+    def disable_motion_detection(self):
+        """ Disable the motion detection settings for the camera.
+            This should be overridden in derived classes """
+        if not self.is_motion_detection_supported:
+            return False
+        if self.is_motion_detection_enabled:
+            return True
+
+    def set_ftp_details(self):
+        """ A stub method that should be overridden in base classes
+        to set the FTP configuration values for the specific device. """
+        if not self.is_motion_detection_supported:
+            return False
+
+    def process_file_event(self, event):
+        """ Event handler for a new FTP upload event from the FTP component.
+            It checks the new file's path against the configured FTP directory
+            to determine whether or not to process the file """
+        if self.ftp_path is not None:
+            if event.data.get('file_name').startswith(self.ftp_path):
+                if self._is_detecting_motion is False:
+                    self._is_detecting_motion = True
+                    self.update_ha_state()
+
+                self._last_motion_detected = datetime.datetime.now()
+                self.process_new_file(event.data.get('file_name'))
+
+    def process_new_file(self, path):
+        """ Used to process a new motion capture image, it finds the most
+        recently created event directory and moves the file there in order
+        to group all images captured as part of the same event together.
+        If the image is the first in the sequence a new directory will be
+        created. """
+        if not os.path.isfile(path):
+            return False
+
+        if self.event_images_path is None:
+            return False
+
+        if not os.path.isdir(self.event_images_path):
+            os.makedirs(self.event_images_path)
+
+        all_subdirs = [
+            d for d in os.listdir(self.event_images_path)
```

Comment:
Also, you might want to consider using the [`pathlib`](https://docs.python.org/3.4/library/pathlib.html#module-pathlib) module:

```python
from pathlib import Path

event_images_dir = Path(self.event_images_path)
all_subdirs = [d for d in p.iterdir() if d.is_dir()]
```